src/Microsoft.Azure.WebJobs.Host/Config/FluentBindingRule.cs

```diff
@@ -167,20 +201,133 @@ public void BindToInput<TType>(IAsyncConverter<TAttribute, TType> builderInstanc
         /// <typeparam name="TType"></typeparam>
         /// <param name="builder"></param>
         /// <returns></returns>
-        public void BindToInput<TType>(Func<TAttribute, TType> builder)
+        public FluentBindToInputOptions BindToInput<TType>(Func<TAttribute, ValueBindingContext, Task<TType>> builder)        
         {
-            var builderInstance = new DelegateConverterBuilder<TAttribute, TType> { BuildFromAttribute = builder };
-            this.BindToInput<TType>(builderInstance);
+            var pm = PatternMatcher.New(builder);
+            return BindToInput<TType>(pm);
+        }
+
+        /// <summary>
+        /// Bind an attribute to the given input, using the supplied delegate to build the input from an resolved 
+        /// instance of the attribute. 
+        /// </summary>
+        /// <typeparam name="TType"></typeparam>
+        /// <param name="builder"></param>
+        /// <returns></returns>
+        public FluentBindToInputOptions BindToInput<TType>(Func<TAttribute, CancellationToken, Task<TType>> builder)
+        {
+            var pm = PatternMatcher.New(builder);
+            return BindToInput<TType>(pm);
+        }
+
+        /// <summary>
+        /// Bind an attribute to the given input, using the supplied delegate to build the input from an resolved 
+        /// instance of the attribute. 
+        /// </summary>
+        /// <typeparam name="TType"></typeparam>
+        /// <param name="builder"></param>
+        /// <returns></returns>
+        public FluentBindToInputOptions BindToInput<TType>(Func<TAttribute, TType> builder)
+        {
+            var pm = PatternMatcher.New(builder);
+            return BindToInput<TType>(pm);
+        }
+
+        // Common worker for BindToInput rules. 
+        private FluentBindToInputOptions BindToInput<TType>(PatternMatcher pm)
+        {
+            var rule = new BindToInputBindingProvider<TAttribute, TType>(_parent.NameResolver, _parent.ConverterManager, pm);
+            return Bind(rule);
         }
 
         #endregion // BindToInput
 
+        #region BindToStream
+
+        /// <summary>
+        /// Bind an attribute to a stream. This ensures the stream is flushed after the user function returns. 
+        /// It uses the attribute's Access property to determine direction (Read/Write). 
+        /// It includes rules for additional types of TextReader,string, byte[], and TextWriter,out string, out byte[].
+        /// </summary>
+        /// <param name="func"></param>
+        /// <param name="fileAccess"></param>
+        public void BindToStream(Func<TAttribute, ValueBindingContext, Task<Stream>> func, FileAccess fileAccess)
```

Comment:
can you rename func? I think you use "builder" in other places? Make consistent throughout